client/consensus/src/collators/lookahead.rs

```diff
@@ -74,8 +84,287 @@ use {
+    OClient: ProvideRuntimeApi<OBlock>
+        + HeaderMetadata<OBlock, Error = sp_blockchain::Error>
+        + HeaderBackend<OBlock>
+        + BlockBackend<OBlock>
+        + BlockIdTo<OBlock>
+        + 'static,
+    OClient::Api: TaggedTransactionQueue<OBlock>
+        + XCMCoreBuyerApi<OBlock, <<OBlock as BlockT>::Header as HeaderT>::Number, ParaId, NimbusId>,
+    TxPool: TransactionPool<Block = OBlock>,
+{
+    // We do nothing if this is not a parathread
+    if aux_data.slot_freq.is_none() {
+        return Err(BuyCoreError::NotAParathread);
+    }
+
+    let orchestrator_best_hash = orchestrator_client.info().best_hash;
+    let orchestrator_runtime_api = orchestrator_client.runtime_api();
+
+    let buy_core_slot_drift = orchestrator_client
+        .runtime_api()
+        .get_buy_core_slot_drift(orchestrator_best_hash)?;
+
+    // Convert drift in terms of container chain slots for parity between client side calculation and
+    // orchestrator runtime calculation
+    // We tried to made this calculation as generic as possible so that it can handle
+    // arbitrary slot timings as well and won't assume anything.
+    // Formula is: (Slot_drift_in_orchestrator_slot * orchestrator_slot_duration) / container_slot_duration
+    let buy_core_container_slot_drift = buy_core_slot_drift
+        .checked_mul(orchestrator_slot_duration.as_millis())
+        .and_then(|intermediate_result| {
+            intermediate_result.checked_div(container_slot_duration.as_millis())
+        })
+        .ok_or(BuyCoreError::SlotDriftConversionOverflow)?;
+
+    let current_container_slot = inherent_providers.slot();
+
+    let slot_claim = tanssi_claim_slot::<P, Block>(
+        aux_data,
+        &parent_header,
+        current_container_slot,
+        ClaimMode::ParathreadCoreBuying {
+            drift_permitted: buy_core_container_slot_drift.into(),
+        },
+        keystore,
+    )
+    .ok_or(BuyCoreError::UnableToClaimSlot)?;
+
+    let pubkey = slot_claim.author_pub;
+
+    orchestrator_runtime_api.is_core_buying_allowed(
+        orchestrator_best_hash,
+        para_id,
+        pubkey.clone(),
+    )??;
+
+    let nonce =
+        orchestrator_runtime_api.get_buy_core_signature_nonce(orchestrator_best_hash, para_id)?;
+
+    let collator_buy_core_proof =
+        BuyCoreCollatorProof::new_with_keystore(nonce, para_id, pubkey, keystore)?
+            .ok_or(BuyCoreError::UnableToFindKeyForSigning)?;
+
+    let extrinsic = orchestrator_runtime_api.create_buy_core_unsigned_extrinsic(
+        orchestrator_best_hash,
+        para_id,
+        collator_buy_core_proof,
+    )?;
+
+    let mut tx_stream = orchestrator_tx_pool
+        .submit_and_watch(orchestrator_best_hash, TransactionSource::Local, *extrinsic)
+        .await
+        .map_err(BuyCoreError::TxSubmissionError)?
+        .fuse();
+
+    // We wait at max till 3 slots
+    let sleep = tokio::time::sleep(orchestrator_slot_duration.as_duration() * 3);
```

Comment:
Now that I think about it, this timeout here looks wrong. The loop in the `run` function gets triggered every time there is a new relay block. But this `try_to_buy_core` function will wait for up to 3 tanssi slots. So it can get stuck waiting for transaction inclusion, while another collator buys a core.

I guess my main concern is that this will cause us to miss relay block notifications... well actually can't it cause us to try to build blocks on top of an older relay block? Since `maybe_relay_parent_header` comes from a stream, and we don't have any logic to skip stale blocks, if this function sleeps then the new relay blocks will start to pile up.

Anyway, my proposal for `try_to_buy_core`:

* Have state about latest sent transaction, initialized to None
* If latest sent transaction is None, or it was sent more than 3 slots ago:
  * Log warning about why it was not included, send another buy core transaction, set state to Some(current_slot), and return immediately
* Else if transaction was sent less than 3 slots ago:
  * Return immediately, we will check if it's included the next time this function is called
* In the outer loop in the `run` function: if the parathread has a core available, set state of transaction to None

Reply:
Regarding your proposal, I think keeping a local state is not something I would prefer (unless there is no other way of course), as this could complicate things when there more than one collators per parathread and each collator's lookahead loop maintains a state. 

I had similar first iteration of this algorithm was where we are not waiting for the unsigned tx to be included in the block but I wanted bit of certainty that the tx was indeed included in the block before we do anything else. Maybe I can just revert to that and see if it makes any difference. 

Another solution that I am experimenting with is where we have another async loop that will try to buy core whenever it is eligible to do so and since it will run in separate loop, there is less chance of missing relay import notification or any other delay that is currently arising from waiting for tx. 

Will write results of these two approaches here and then we can decide.

Reply:
Keeping a local state is ugly but I think it's the simplest approach, I'm a bit scared of the idea of a separate loop but it could also work so yeah let's try.